modules/contact/src/actions/ContactAction.C

```diff
@@ -1047,6 +1061,149 @@ ContactAction::addNodeFaceContact()
+            { return first_pair.second > second_pair.second; });
+
+  // build kd-tree
+  using KDTreeType = nanoflann::KDTreeSingleIndexAdaptor<
+      nanoflann::L2_Simple_Adaptor<Real, PointListAdaptor<NodeBoundaryIDInfo>, Real, std::size_t>,
+      PointListAdaptor<NodeBoundaryIDInfo>,
+      LIBMESH_DIM,
+      std::size_t>;
+
+  // This parameter can be tuned. Others use '10'
+  const unsigned int max_leaf_size = 20;
+
+  // Build point list adaptor with all nodes-sidesets pairs for possible mechanical contact
+  auto point_list = PointListAdaptor<NodeBoundaryIDInfo>(node_boundary_id_vector.begin(),
+                                                         node_boundary_id_vector.end());
+  auto kd_tree = std::make_unique<KDTreeType>(
+      LIBMESH_DIM, point_list, nanoflann::KDTreeSingleIndexAdaptorParams(max_leaf_size));
+
+  if (!kd_tree)
+    mooseError("Internal error. KDTree was not properly initialized in the contact action.");
+
+  kd_tree->buildIndex();
+
+  // data structures for kd-tree search
+  nanoflann::SearchParams search_params;
+  std::vector<std::pair<std::size_t, Real>> ret_matches;
+
+  // For all nodes
+  for (const auto & pair : node_boundary_id_vector)
+  {
+    // clear result buffer
+    ret_matches.clear();
+
+    // position where we expect a periodic partner for the current node and boundary
+    const Point search_point = *pair.first;
+    const auto radius_for_search = getParam<Real>("automatic_pairing_distance");
+
+    // search at the expected point
+    kd_tree->radiusSearch(
+        &(search_point)(0), radius_for_search * radius_for_search, ret_matches, search_params);
+
+    for (auto & match_pair : ret_matches)
+    {
+      const auto & match = node_boundary_id_vector[match_pair.first];
+      // If the proximity node identified belongs to a boundary in the input, add boundary pair
+
+      // Make sure node is on a boundary chosen for contact mechanics
+      auto it = std::find(_automatic_pairing_boundaries_id.begin(),
+                          _automatic_pairing_boundaries_id.end(),
+                          match.second);
+
+      // If nodes are on the same boundary, pass.
+      if (match.second == pair.second)
+        continue;
+
+      // At this point we will likely create many repeated pairs because many nodal pairs may
+      // fulfill the distance condition imposed by the automatic_pairing_distance user input
+      // parameter.
+      if (it != _automatic_pairing_boundaries_id.end())
+      {
+        const int index_one = it - _automatic_pairing_boundaries_id.begin();
+        auto it_other = std::find(_automatic_pairing_boundaries_id.begin(),
+                                  _automatic_pairing_boundaries_id.end(),
+                                  pair.second);
+
+        if (it_other == _automatic_pairing_boundaries_id.end())
+          mooseError("Error in contact action. Unable to find boundary ID for node proximity "
+                     "automatic pairing.");
```

Comment:
I think this could just be an assertion, but an error is fine too

Reply:
I preferred the error initially. But this should probably never happen, so assertion is good.